paddleslim/prune/prune_walker.py

```diff
@@ -0,0 +1,557 @@
+# Copyright (c) 2019  PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import numpy as np
+from ..core import Registry
+from ..common import get_logger
+
+__all__ = ["PRUNE_WORKER", "conv2d"]
+
+_logger = get_logger(__name__, level=logging.INFO)
+
+PRUNE_WORKER = Registry('prune_worker')
+
+class PruneWorker(object):
+    def __init__(self, op, pruned_params=[], visited={}):
+        self.op = op
+        self.pruned_params = pruned_params
+        self.visited=visited
+
+    def prune(self, var, pruned_aixs, pruned_idx):
+        pass
+
+    def prune_op(self, op, var, pruned_aixs, pruned_idx, visited=None):
+        if op.type().endswith("_grad"):
+            return
+        if visited is not None:
+            self.visited = visited
+        cls = PRUNE_WORKER.get(op.type())
+        assert cls is not None , "The walker of {} is not registered.".format(op.type())
+        _logger.debug("\nfrom: {}\nto: {}\npruned_aixs: {}; var: {}".format(self.op, op, pruned_aixs, var.name()))
+        walker = cls(op, pruned_params=self.pruned_params, visited=self.visited)
+        walker.prune(var, pruned_aixs, pruned_idx)
+
+@PRUNE_WORKER.register
+class conv2d(PruneWorker):
+
+    def __init__(self, op, pruned_params, visited={}):
+        super(conv2d, self).__init__(op, pruned_params, visited)
+
+    def prune(self, var, pruned_axis, pruned_idx):
+        key = "_".join([str(self.op.idx()), var.name()])
+        if pruned_axis not in self.visited:
+            self.visited[pruned_axis] = {}
+        if key in self.visited[pruned_axis]:
+            return
+        else:
+           self.visited[pruned_axis][key] = True
```

Comment:
line 53 to line 59 are repeated in all subclass, maybe create a common fucntion in superclass